ent/encryption/encrypted_file_impl.cc

```diff
@@ -701,5 +701,56 @@ std::unique_ptr<data_sink_impl> make_encrypted_sink(data_sink sink, shared_ptr<s
     return std::make_unique<encrypted_data_sink>(std::move(sink), std::move(k));
 }
 
+class encrypted_data_source : public data_source_impl, public block_encryption_base {
+    data_source _source;
+    size_t current_position = 0;
+
+public:
+    encrypted_data_source(data_source source, shared_ptr<symmetric_key> k) : block_encryption_base(std::move(k)), _source(std::move(source)) {}
+
+    future<temporary_buffer<char>> get() override {
+        return _source.get().then([this](temporary_buffer<char> buf) {
+            if (buf.empty()) {
+                return buf;
+            }
+
+            const size_t len = buf.size();
+            temporary_buffer<char> output(len);
+            const size_t key_block_size = _key->block_size();
+            size_t decrypted_bytes = 0;
+
+            for (size_t offset = 0; offset < len; offset += block_size) {
+                auto iv = iv_for(current_position + offset);
+                // Determine how many bytes remain in this block
+                const size_t remaining = std::min<uint64_t>(block_size, len - offset);
+
+                // If the remaining bytes form a partial block, align down to a whole block size
+                size_t bytes_to_decrypt = (remaining < block_size) ? align_down(remaining, key_block_size) : block_size;
```

Comment:
This assumes that we get data from underlying source in at least block sizes >= our own block size, which might not be true for a disk source, IIRC (we use 4096, pretty high, a file might give us 512b chunks). 

More to the point, a network source might give us wholly unaligned data (is that valid?). 
As much as it pains me, I think you might need to use buffering here.

Reply:
hmmm... not sure I'm following
if I get `buf.size()` smaller than `block_size` which is 4kib then I align to `_key->block_size()` and then decrypt it. Isnt it supposed to cover all cases?

Reply:
Or to put it the other way, in your opinion, [here](https://github.com/scylladb/scylladb/pull/23695/files#diff-19c2c9cd31d8f6d9cdcaa35650bb9ab48e97f3943c7b932d17e598ce72a8ef56R545), what case could possibly reproduce the problem you have in mind

Reply:
It does, but the problems are: 
a.) iv_for assumes (and requires) the position given to be at (data) block offset.
b.) What happens with the data remaining? Unless you are guaranteed to get aligned blocks from underlying source (which again, with for example a network source I am not sure we can assume), you might end up with trailing data that needs to be kept until you actually get an empty (EOS) return from said source.

Reply:
done